src/cpu/flexcpu/flexcpu_thread.cc

```diff
@@ -652,9 +750,76 @@ FlexCPUThread::onExecutionCompleted(shared_ptr<InflightInst> inst_ptr,
     if (inst_ptr->staticInst()->isControl()) {
         // We must have our next PC now, since this branch has just resolved
 
-        // If we are a control instruction, then we were unable to begin a
-        // fetch during decode, so we should start one now.
-        advanceInst(getNextPC());
+        if (_cpuPtr->hasBranchPredictor()) {
+            // If a branch predictor has been set, then we need to check if
+            // a prediction was made, squash instructions if incorrectly
+            // predicted, and notify the predictor accordingly.
+
+            auto it = find(inflightInsts.begin(), inflightInsts.end(),
+                           inst_ptr);
+
+            if (it == inflightInsts.end()) {
+                // Should be unreachable...
+                panic("Did a completion callback on an instruction not in "
+                      "our buffer.");
+            }
+
+            shared_ptr<InflightInst> following_inst =
```

Comment:
How about if next iterator (whatever that looks like in C++) is not the end or `advanceInstNeedsRetry`. Then you don't have to have the null check.

Reply:
I don't like the way the code looks if we separate out the iterator access that way. More lines, and we still have to have the null check again about 7 lines later.